CART update

Saurabh7 opened the pull request.

Saurabh7 commented:
Some values on datasets: 
old : https://gist.github.com/Saurabh7/e724f17b3d0558efbf7bc434b089a904
new : https://gist.github.com/Saurabh7/87d53f12f04670e5f64db22c30bb107d

karlnapf commented:
Nice changes and useful. But man, this takes like 10 mins to do. Where is the other progress? :)

Saurabh7 commented:
Updated with few things.
Removed unnecessary loops, integrated with another one.
Changed the way we do argsort. I tried a few things here. New one should be much more cache friendly. Found this a bit hard to track down, as sorting is already expensive. another mayb small thing but looks like has pretty good impact on performance :)

Need to double check few things with regression, extend this to the other case of nominal attributes if looks good and refactor a bit.

Saurabh7 commented:
latest output: https://gist.github.com/Saurabh7/f7ced9fbdcf50eb6fe555a8f6acb618a

Adding this to benchmark framework https://github.com/zoq/benchmarks/pull/16

karlnapf commented on src/shogun/multiclass/tree/RandomCARTree.cpp:
```diff
@@ -108,18 +108,20 @@ int32_t CRandomCARTree::compute_best_attribute(const SGMatrix<float64_t>& mat, c
 
 		// number of non-missing vecs
 		int32_t n_nm_vecs=feats.vlen;
-		while (feats[sorted_args[n_nm_vecs-1]]==MISSING)
```
whats the value of MISSING?

Saurabh7 replied:
its set to `CMath::MAX_REAL_NUMBER` , an optional routine in case missing attribute values

karlnapf replied:
why not set it to NAN?

karlnapf commented:
Lets aim to merge this soon, before the parallel stuff starts.
Just ping me and squash to trigger that

karlnapf commented on src/shogun/multiclass/tree/RandomCARTree.cpp:
```diff
@@ -49,238 +49,22 @@ void CRandomCARTree::set_feature_subset_size(int32_t size)
-	" or equal to the total number of features(%d here)\n",m_randsubset_size,num_feats)
+{
+	REQUIRE(m_randsubset_size<=mat.num_rows, "The Feature subset size(set %d) should be less than"
+	" or equal to the total number of features(%d here)\n",m_randsubset_size,mat.num_rows)
```
remove "here"
Also, please make really clear what things are here. Are you talking about subsets of data or of components?
Thirdly, is the user able to cause this error message?

karlnapf commented on src/shogun/multiclass/tree/CARTree.h:
```diff
@@ -421,6 +421,9 @@ class CCARTree : public CTreeMachine<CARTreeNodeData>
 	SGVector<bool> m_nominal;
 
+	/** Vector to keep track of constant features to skip */
+	SGVector<bool> m_constant_feats;
```
Can you ellaborate a bit here?

karlnapf commented on src/shogun/multiclass/tree/CARTree.cpp:
```diff
@@ -517,6 +520,14 @@ int32_t CCARTree::compute_best_attribute(SGMatrix<float64_t> mat, SGVector<float
+	if (subset_size)
+	{
+		num_feats=subset_size;
+		CMath::permute(idx);
```
Note that if you permute here and then later loop over idx, then your memory access will be random.

karlnapf commented on src/shogun/machine/RandomForest.h:
```diff
@@ -154,6 +156,8 @@ class CRandomForest : public CBaggingMachine
 	/** weights */
 	SGVector<float64_t> m_weights;
 
+	SGMatrix<float64_t> m_sorted_feats;
```
m_sorted_transposed_feats maybe?

karlnapf commented on src/shogun/multiclass/tree/CARTree.cpp:
```diff
@@ -285,6 +285,29 @@ bool CCARTree::train_machine(CFeatures* data)
+{
+	for(int32_t i=0; i<sorted_indices.num_cols; i++)
+		for(int32_t j=0; j<sorted_indices.num_rows; j++)
+			sorted_indices(j,i)=j;
```
memset!

Saurabh7 replied:
Couldn't find a way to memset this as we need to set rows to constant (not continuous in memory)

karlnapf replied:
ah ok!

karlnapf replied:
Actually, so the inner loop definitely can go, since it accesses a column of the indices.

karlnapf commented on src/shogun/multiclass/tree/CARTree.cpp:
```diff
@@ -285,6 +285,29 @@ bool CCARTree::train_machine(CFeatures* data)
+		for(int32_t j=0; j<sorted_indices.num_rows; j++)
+			sorted_indices(j,i)=j;
+
+	Map<MatrixXd> map_sorted_feats(sorted_feats.matrix, data.num_cols, data.num_rows);
```
is this NULL at this point? It should be

karlnapf commented on src/shogun/multiclass/tree/CARTree.cpp:
```diff
@@ -285,6 +285,29 @@ bool CCARTree::train_machine(CFeatures* data)
+
+	map_sorted_feats=map_data.transpose();
+
+	for(int32_t i=0; i<sorted_feats.num_cols; i++)
```
parallel! :)

karlnapf commented on src/shogun/multiclass/tree/CARTree.cpp:
```diff
@@ -532,17 +571,62 @@ int32_t CCARTree::compute_best_attribute(const SGMatrix<float64_t>& mat, const S
+		for(int32_t j=0;j<active_indices.size();j++)
+		{
+			if (indices_mask[active_indices[j]]>=0)
+				dupes[indices_mask[active_indices[j]]]=j;
```
minor: newline after this pls

karlnapf commented on src/shogun/multiclass/tree/CARTree.cpp:
```diff
@@ -532,17 +571,62 @@ int32_t CCARTree::compute_best_attribute(const SGMatrix<float64_t>& mat, const S
+	{
+		indices_mask = SGVector<int32_t>(mat.num_rows);
+		indices_mask.set_const(-1);
+		for(int32_t j=0;j<active_indices.size();j++)
```
I hope active_indices is sorted?

karlnapf commented on src/shogun/multiclass/tree/CARTree.cpp:
```diff
@@ -532,17 +571,62 @@ int32_t CCARTree::compute_best_attribute(const SGMatrix<float64_t>& mat, const S
+			SGVector<float64_t> temp_vec(mat.get_column_vector(idx[i]), mat.num_rows, false);
+			SGVector<int32_t> sorted_indices(m_sorted_indices.get_column_vector(idx[i]), mat.num_rows, false);
+			int32_t count=0;
+			for(int32_t j=0;j<mat.num_rows;j++)
```
wait, why are you still traversing row wise?
This outerloop should be over columns

Saurabh7 replied:
Inside a column here, needed to check which indices are active.

karlnapf commented on src/shogun/multiclass/tree/CARTree.cpp:
```diff
@@ -611,7 +695,12 @@ int32_t CCARTree::compute_best_attribute(const SGMatrix<float64_t>& mat, const S
-
+				for (int32_t j=n_nm_vecs-1;j>=0;j--)
+				{
+					if(dupes[j]!=j)
```
dupes is sorted?

Saurabh7 replied:
yes

vigsterkr commented on src/shogun/machine/RandomForest.cpp:
```diff
@@ -159,11 +159,21 @@ void CRandomForest::set_machine_parameters(CMachine* m, SGVector<index_t> idx)
 
+bool CRandomForest::train_machine(CFeatures* data)
+{
+	SGMatrix<float64_t> mat=(dynamic_cast<CDenseFeatures<float64_t>*>(data))->get_feature_matrix();
```
hardcoding that a random forest can only work with `CDenseFeatures<float64_t>`
which in case of a decision tree is very very crippling effect as they should handle categorical values *easily*....

Saurabh7 replied:
The current way gets information about feature types (nominal or not) through a parameter vector and handles each feature later. But yeah the supported type is only densefeature.

vigsterkr replied:
i'm not sure if i understand. `get_feature_type()` and `get_feature_class()` functions will give you the exact type of features you are dealing with.
this should be a rather easy patch to add support for other types than `DenseFeatures`

Saurabh7 replied:
What I meant was cases like feature1 is nominal , feature2 is continuous.
This assumes that string values like 'hot', 'cold' are already encoded in numerals. 
So the idea is we should support string features and do this conversion internally ?

karlnapf replied:
There is a couple of options here. But this PR is about speeding the *existing* implementation up.
So lets merge this while it is small, and then open a new one for the ordinals

vigsterkr replied:
yep yep we've talked about this with @Saurabh7 
we should finish up with this PR as a performance fix, and then work with further improving trees/forests.

vigsterkr commented on src/shogun/multiclass/tree/CARTree.cpp:
```diff
@@ -228,7 +228,7 @@ void CCARTree::set_max_depth(int32_t depth)
 {
 	return m_min_node_size;
-}
+}	
```
why this extra whitespace?

vigsterkr commented:
one more thing. please make sure that the following error is not in the updated code:
http://buildbot.shogun-toolbox.org/static_analysis/2016-06-26-005840-32416-1/report-9779d7.html#EndPath

karlnapf commented:
@vigsterkr Shouldnt we fix these in a separate patch to keep this one smaller? (It is only about optimising speed with the existing implementation). But definitely should fix the static analysis errors

karlnapf commented on src/shogun/multiclass/tree/CARTree.cpp:
```diff
@@ -230,7 +230,7 @@ void CCARTree::set_max_depth(int32_t depth)
 int32_t CCARTree::get_min_node_size() const
 {
 	return m_min_node_size;
-}	
```
try to avoid such whitespace changes

karlnapf commented on src/shogun/machine/RandomForest.cpp:
```diff
@@ -159,11 +159,26 @@ void CRandomForest::set_machine_parameters(CMachine* m, SGVector<index_t> idx)
+	if (data == NULL)
+		data = m_features;
+	
+	REQUIRE(data, "Training features not set!");
```
needs newline at the end

karlnapf commented on src/shogun/machine/RandomForest.h:
```diff
@@ -154,6 +156,11 @@ class CRandomForest : public CBaggingMachine
+	SGMatrix<float64_t> m_sorted_transposed_feats;
+
+	/** Indices of pre-sorted features */
+	SGMatrix<int32_t> m_sorted_indices;
```
index_t

karlnapf commented on src/shogun/multiclass/tree/CARTree.cpp:
```diff
@@ -102,6 +105,8 @@ CMulticlassLabels* CCARTree::apply_multiclass(CFeatures* data)
 	// apply multiclass starting from root
 	bnode_t* current=dynamic_cast<bnode_t*>(get_root());
+	
+	REQUIRE(current, "Tree root not present, train tree machine before apply\n");
```
Minor. Period at end of sentence

karlnapf replied:
And remove this "Tree root not present. Just say that machine not yet trained"

karlnapf commented on src/shogun/multiclass/tree/CARTree.cpp:
```diff
@@ -1235,6 +1366,8 @@ float64_t CCARTree::compute_error(CLabels* labels, CLabels* reference, SGVector<
 
 CDynamicObjectArray* CCARTree::prune_tree(CTreeMachine<CARTreeNodeData>* tree)
 {
+	REQUIRE(tree, "Tree not provided for pruning\n");	
```
Minor. Period after sentence.
Can the user cause this?

Saurabh7 replied:
No, this is protected method.

karlnapf commented on src/shogun/multiclass/tree/RandomCARTree.cpp:
```diff
@@ -49,238 +49,27 @@ void CRandomCARTree::set_feature_subset_size(int32_t size)
+		m_randsubset_size=CMath::sqrt(mat.num_rows-0.f);
+	subset_size=m_randsubset_size;
+	
+	REQUIRE(subset_size<=num_feats, "The Feature subset size(set %d) should be less than"
```
period

karlnapf commented:
Fine to merge from my side, just some very minor comments

karlnapf commented:
@vigsterkr check and feel free to merge

Saurabh7 commented:
Outputs : https://gist.github.com/Saurabh7/0964969665bdede7934f4cef5b2a8ab9
script: https://gist.github.com/Saurabh7/ca352df35a52768c4891fcbdba5f6406

vigsterkr commented:
@Saurabh7 great work! i'd happily merge this but i prefer to have it as one commit (i.e. could you please squash the commits into one) as well as address those minor comments from @karlnapf 
plz ping me when it's done. thnx!

Saurabh7 commented:
cool ! I have updated comments , squashed and fixed memleaks caused due to not freeing the Subset-stack.

vigsterkr commented:
@Saurabh7 great! i'll wait for travis to finish and then i'll merge it! great work!
